lib/Dialect/TTIR/Transforms/FlattenSlidingWindow.cpp

```
@@ -0,0 +1,154 @@
+        inputType.getDimSize(2));
+
+    auto newPool = ttmlir::utils::createDPSOp<ttir::MaxPool2dOp>(
+        rewriter, op.getLoc(), getNHWFlattenedType(outputType), flattenedInput,
+        adaptor.getKernelHeight(), adaptor.getKernelWidth(),
+        adaptor.getStrideHeight(), adaptor.getStrideWidth(),
+        adaptor.getDilationHeight(), adaptor.getDilationWidth(),
+        adaptor.getCeilMode(), adaptor.getPaddingLeft(),
+        adaptor.getPaddingRight(), adaptor.getPaddingTop(),
+        adaptor.getPaddingBottom(), flattenedCompatInfoAttr);
+
+    Value output = generateReshape(newPool, outputType, rewriter);
+
+    rewriter.replaceOp(op, output);
+    return success();
+  }
+};
+} // namespace
+
+class TTIRFlattenSlidingWindow
+    : public impl::TTIRFlattenSlidingWindowBase<TTIRFlattenSlidingWindow> {
+public:
+  using impl::TTIRFlattenSlidingWindowBase<
+      TTIRFlattenSlidingWindow>::TTIRFlattenSlidingWindowBase;
+
+  void runOnOperation() final {
+    RewritePatternSet conversionPatterns(&getContext());
+    TypeConverter typeConverter;
+    // All types map 1:1.
+    typeConverter.addConversion([](Type type) { return type; });
+    conversionPatterns
+        .add<ConvertToFlattenedConv2dPattern,
+             ConvertToMaxPool2dFlattenedCompatOpConversionPattern>(
+            typeConverter, &getContext());
+    FrozenRewritePatternSet conversionPatternSet(std::move(conversionPatterns));
+
+    mlir::ConversionTarget target(getContext());
+    target.addLegalDialect<ttir::TTIRDialect>();
+    target.addLegalDialect<mlir::func::FuncDialect>();
+
+    target.addDynamicallyLegalOp<ttir::Conv2dOp>([&](ttir::Conv2dOp op) {
+      return op.getFlattenedCompatInfo().has_value();
+    });
+    target.addDynamicallyLegalOp<ttir::MaxPool2dOp>([&](ttir::MaxPool2dOp op) {
+      return op.getFlattenedCompatInfo().has_value();
+    });
+
+    if (failed(applyPartialConversion(getOperation(), target,
```

I am curious: why does this pass choose to use the dialect conversion driver instead of the greedy driver?

---

I need to make certain ops illegal. No sliding window op should be unflattened by the end of this pass. If these patterns are edited in the future, and for one reason or antother this causes a sliding window op to _not_ be flattened, we need to raise an error.

The reason I used `applyPartialConversion` is because it allows me to pass `target` to define what is illegal and what is legal. I don't believe the greedy driver accepts a `ConversionTarget`

---

I understand. There is a certain level of convenience to being able to specify what's legal/illegal for a pass target, a kind of post-invariant.

I bring this up because the dialect conversion framework is in a state of flux right now within the parent MLIR project, due to issues with type conversion materialization and costs of DAG cloning while supporting rollback. (Of course, these type conversion issues don't show up for our passes that don't actually need to convert types.)

> No sliding window op should be unflattened by the end of this pass.

Note that technically the greedy driver can accomplish the same invariant when used with patterns that return `failure()` from `matchAndRewrite()` for ops that have been flattened. Both drivers use a worklist impl and mutated /new ops are always revisited again as the driver checks for its fixed point.

Anyway, I just wanted to understand the motivation, thanks for responding.